backend/src/main/java/de/bund/digitalservice/ris/norms/application/service/LdmlDeValidator.java

```diff
@@ -85,4 +99,81 @@ public void fatalError(SAXParseException exception) {
 
     return Norm.builder().document(result).build();
   }
+
+  /**
+   * Validate the norm against the schematron rules. Throws if a rule is not fulfilled.
+   *
+   * @param norm the norm to validate
+   * @throws IOException The schematron file couldn't be read
+   * @throws XmlProcessingException A problem occurred during the processing of the XML
+   * @throws LdmlDeSchematronException A Schematron rule is violated
+   */
+  public void validateSchematron(Norm norm) throws IOException {
+    Source xsltSource = new StreamSource(schematronXslt.getInputStream());
+    xsltSource.setSystemId(schematronXslt.getURL().toString());
+    Transformer transformer = null;
+    try {
+      transformer = new TransformerFactoryImpl().newTransformer(xsltSource);
+    } catch (TransformerConfigurationException e) {
+      throw new XmlProcessingException(e.getMessage(), e);
+    }
+
+    String xmlText = XmlMapper.toString(norm.getDocument());
+    Source xmlSource = new StreamSource(new ByteArrayInputStream(xmlText.getBytes()));
+
+    var result = new DOMResult();
+    try {
+      transformer.transform(xmlSource, result);
+    } catch (TransformerException e) {
+      throw new XmlProcessingException(e.getMessage(), e);
+    }
+
+    if (!(result.getNode() instanceof Document resultDoc)) {
+      throw new XmlProcessingException("Result is not a document.", null);
+    }
+
+    var failedAsserts =
+        resultDoc.getElementsByTagNameNS("http://purl.oclc.org/dsdl/svrl", "failed-assert");
+    // successful-reports are often used for warnings, and do not indicate that an assert was not
+    // failed
+    var successfulReports =
+        resultDoc.getElementsByTagNameNS("http://purl.oclc.org/dsdl/svrl", "successful-report");
+
+    if (failedAsserts.getLength() > 0 || successfulReports.getLength() > 0) {
+      var failedAssertMessages = NodeParser.nodeListToList(failedAsserts).stream();
+      var successfulReportMessages = NodeParser.nodeListToList(successfulReports).stream();
+
+      var errors =
+          Stream.concat(failedAssertMessages, successfulReportMessages)
+              .map(
+                  node -> {
+                    // The location includes an XPath with expanded QNames
+                    // (Q{namespace}<localPart>).
+                    var xPath = node.getAttributes().getNamedItem("location").getNodeValue();
+
+                    // Find the eId of the node responsible for this problem. Sometimes the location
+                    // points to an attribute, so we might need to move up in the element tree to
+                    // find the xPath.
+                    List<Node> eIdNodes =
```

Comment:
Should we maybe cache the xpath results? So that we don't need to compute it again if we are querying the same xpath, since xpath evaluation is relatively expensive and the catching can be easily achieved with:

```
                          List<Node> eIdNodes = xPathCache.computeIfAbsent(
                              xPath + "/ancestor-or-self::*/@eId",
                              key -> NodeParser.getNodesFromExpression(key, norm.getDocument())
                          );
```